core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -341,9 +341,11 @@ class FileMessageSet private[kafka](@volatile var file: File,
     if(targetSize > originalSize || targetSize < 0)
       throw new KafkaException("Attempt to truncate log segment to " + targetSize + " bytes failed, " +
                                " size of this log segment is " + originalSize + " bytes.")
-    channel.truncate(targetSize)
-    channel.position(targetSize)
-    _size.set(targetSize)
+    if (targetSize != channel.size().toInt) {
```

Comment:
Thank you @ijuma and @junrao  for the code review. I pushed another commit with the latest improvements and added a testcase for targetSize > originalSize.

The Jenkins build failed on the last few commits, but I think this unrelated. The project builds fine locally.

Is there anything more I can provide for this PR?